app/src/test/java/io/hstream/testing/ScaleTest.java

```diff
@@ -0,0 +1,297 @@
+    hStreamClient.close();
+  }
+
+  // -----------------------------------------------------------------------------------------------
+
+  @RepeatedTest(10)
+  @Timeout(120)
+  void testLargeNumProducer() throws Exception {
+    final String stream = randStream(hStreamClient);
+    final String subscription = randSubscription(hStreamClient, stream);
+    final int total = 64;
+
+    List<RecordId> recordIds0 = new ArrayList<>();
+    ReentrantLock lock = new ReentrantLock();
+
+    List<Thread> threads = new ArrayList<>();
+    for (int i = 0; i < total; ++i) {
+      Thread thread =
+          new Thread(
+              () -> {
+                Producer producer = hStreamClient.newProducer().stream(stream).build();
+                for (int j = 0; j < total; ++j) {
+                  lock.lock();
+                  try {
+                    recordIds0.add(producer.write(randBytes()).join());
+                  } catch (Throwable e) {
+                    logger.info("========e {}", e.getMessage());
+                    e.printStackTrace();
+                    recordIds0.add(producer.write(randBytes()).join());
+                  }
+                  lock.unlock();
+                }
+              });
+      thread.start();
+      threads.add(thread);
+    }
+    for (var x : threads) {
+      x.join();
+    }
+
+    Set<RecordId> set0 = new HashSet<>(recordIds0);
+    Set<RecordId> set1 = new HashSet<>();
+    CountDownLatch countDown1 = new CountDownLatch(total * total);
+    Consumer consumer =
+        hStreamClient
+            .newConsumer()
+            .subscription(subscription)
+            .rawRecordReceiver(
+                (recs, recv) -> {
+                  recv.ack();
+                  if (set1.add(recs.getRecordId())) {
+                    countDown1.countDown();
+                  }
+                })
+            .build();
+
+    consumer.startAsync().awaitRunning();
+    Assertions.assertTrue(countDown1.await(90, TimeUnit.SECONDS));
+    consumer.stopAsync().awaitTerminated();
+
+    Assertions.assertEquals(set0, set1);
+    Assertions.assertEquals(total * total, recordIds0.size());
+  }
+
+  @Test
+  @Timeout(120)
+  void testLargeNumConsumer() throws Exception {
+    final String stream = randStream(hStreamClient);
+    final String subscription = randSubscription(hStreamClient, stream);
+    final int total = 64;
+
+    Producer producer = hStreamClient.newProducer().stream(stream).build();
+    Set<RecordId> recordIds0 = new HashSet<>();
+    for (int i = 0; i < total; ++i) {
+      recordIds0.add(
+          producer.write(UUID.randomUUID().toString().getBytes(StandardCharsets.UTF_8)).join());
+    }
+
+    Thread cur = Thread.currentThread();
+
+    Set<RecordId> recordIds1 = new HashSet<>();
+    ReentrantLock lock = new ReentrantLock();
+    CountDownLatch countDown = new CountDownLatch(total);
+
+    List<Thread> threads = new ArrayList<>();
+    for (int i = 0; i < total; ++i) {
+      Thread thread =
```

Comment:
I am a little confused by the threads here... By `consumer.startAsync()`, a consumer will run its tasks asyncally(in another thread, or in many different threads from a pool) and when an exception happens it will be thrown to the topmost. So I do not know what is the advantage over something like
```
for (...) {
    Consumer consumer = ...;
    consumers.add(consumer);
    consumer.startAsync()...
    ...
}
...
```

Reply:
that's nicer, fixed